src/psyclone/psyir/transformations/intrinsics/dotproduct2code_trans.py

```diff
@@ -0,0 +1,321 @@
+            slice notation and is not a 1d array.
+        :raises TransformationError: if an argument uses array slice \
+            notation but the array slice is not for the first dimension of \
+            the array.
+        :raises TransformationError: if an argument uses array slice \
+            notation but it is not for the full range of the dimension.
+
+        '''
+        super().validate(node, options)
+
+        # Both arguments should be references (or array references)
+        for arg in node.children:
+            if not isinstance(arg, Reference):
+                raise TransformationError(
+                    f"The DotProduct2CodeTrans transformation only supports "
+                    f"the transformation of a dotproduct intrinsic if its "
+                    f"arguments are arrays, but found {self._writer(arg)} in "
+                    f"{self._writer(node)}.")
+
+        for arg in node.children:
+            # The argument should be a 1D array if the argument does
+            # not provide any array slice information (i.e. it is a
+            # Reference)
+            if arg.__class__ is Reference:
+                symbol = arg.symbol
+                # This symbol should be a 1D array
+                if not(isinstance(symbol, DataSymbol) and symbol.is_array and
+                       len(symbol.shape) == 1):
+                    raise TransformationError(
+                        f"The DotProduct2CodeTrans transformation only "
+                        f"supports the transformation of a dotproduct "
+                        f"intrinsic with an argument not containing an array "
+                        f"slice if the argument is a 1D array, but found "
+                        f"{self._writer(arg)} in {self._writer(node)}.")
+
+        for arg in node.children:
+            # If the argument does provide array slice information
+            # then check the array slice is in the first dimension of
+            # the array and that the slice is for the full range of that
+            # dimension i.e. uses a ':'.
+            if isinstance(arg, ArrayReference):
+                if not isinstance(arg.indices[0], Range):
+                    raise TransformationError(
+                        f"The DotProduct2CodeTrans transformation only "
+                        f"supports the transformation of a dotproduct "
+                        f"intrinsic with an argument containing an array "
+                        f"slice if the array slice is for the 1st dimension "
+                        f"of the array, but found {self._writer(arg)} in "
+                        f"{self._writer(node)}.")
+
+                if not arg.is_full_range(0):
+                    raise TransformationError(
+                        f"The DotProduct2CodeTrans transformation only "
+                        f"supports the transformation of a dotproduct "
+                        f"intrinsic with an argument not an array "
```

Comment:
"...an argument containing an array slice if the..."

Reply:
Oops.